### ISSUE TITLE
pendulum object has no attribute 'translate'. 

### ISSUE THREAD

**Issue by lijoev** [sdispater/pendulum#267]
I am using apache airflow 1.10 and the insallation of airflow also installed pendulam. So while running the job it throws an error like pendulum object has no attribute 'translate'.  i am using mysql

**Comment by sdispater** [comment]
Which version of pendulum do you have? And can you give me the complete stack trace?

**Comment by lijoev** [comment]
Hi @sdispater please find the complete Stack trace below
`Traceback (most recent call last): File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app response = self.full_dispatch_request() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request rv = self.handle_user_exception(e) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception reraise(exc_type, exc_value, tb) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise raise value File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request rv = self.dispatch_request() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request return self.view_functions[rule.endpoint](**req.view_args) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask_admin/base.py", line 69, in inner return self._run_view(f, *args, **kwargs) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask_admin/base.py", line 368, in _run_view return fn(self, *args, **kwargs) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/flask_login.py", line 755, in decorated_view return func(*args, **kwargs) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/airflow/www/utils.py", line 268, in wrapper session.commit() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 927, in commit self.transaction.commit() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 467, in commit self._prepare_impl() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 447, in _prepare_impl self.session.flush() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2209, in flush self._flush(objects) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2329, in _flush transaction.rollback(_capture_exception=True) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__ compat.reraise(exc_type, exc_value, exc_tb) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 187, in reraise raise value File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2293, in _flush flush_context.execute() File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 389, in execute rec.execute(self) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 548, in execute uow File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 181, in save_obj mapper, table, insert) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 835, in _emit_insert_statements execute(statement, params) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 945, in execute return meth(self, multiparams, params) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection return connection._execute_clauseelement(self, multiparams, params) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement compiled_sql, distilled_params File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context context) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1405, in _handle_dbapi_exception util.reraise(*exc_info) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 187, in reraise raise value File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context context) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute cursor.execute(statement, parameters) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/cursors.py", line 168, in execute query = self.mogrify(query, args) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/cursors.py", line 147, in mogrify query = query % self._escape_args(args, conn) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/cursors.py", line 127, in _escape_args return dict((key, conn.literal(val)) for (key, val) in args.items()) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/cursors.py", line 127, in <genexpr> return dict((key, conn.literal(val)) for (key, val) in args.items()) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/connections.py", line 469, in literal return self.escape(obj, self.encoders) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/connections.py", line 462, in escape return converters.escape_item(obj, self.charset, mapping=mapping) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/converters.py", line 27, in escape_item val = encoder(val, mapping) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/converters.py", line 118, in escape_unicode return u"'%s'" % _escape_unicode(value) File "/home/nelson/workspace/knowenv/lib/python3.6/site-packages/pymysql/converters.py", line 73, in _escape_unicode return value.translate(_escape_table) AttributeError: 'Pendulum' object has no attribute 'translate'`

**Comment by sdispater** [comment]
Which version of pendulum do you have? Airflow is only compatible with Pendulum < 2.0 so that might be the issue.

**Comment by nrhvyc** [comment]
I also received this error using Airflow 1.10.0 with pymysql and pendulum. Airflow 1.10 uses pendulum 1.4.4. I had to switch to mysqldb to fix this.

**Comment by sdispater** [comment]
Does airflow properly registers the `Pendulum` type as detailed here https://github.com/sdispater/pendulum/tree/1.x#limitations?

If not you might have to do it yourself.

**Comment by lijoev** [comment]
I could fix it using mysqldb.  But didn't know why it comes for pymsql.  Thanks @sdispater

**Comment by cdmikechen** [comment]
@lijoev @sdispater @nrhvyc 
as sdispater said in https://github.com/sdispater/pendulum/tree/1.x#limitations, if use airflow 1.10 with python3, we need to change `airflow/settings.py` in 208-211 rows like this:
```
207     try:
208         #import MySQLdb.converters
209         #MySQLdb.converters.conversions[Pendulum] = MySQLdb.converters.DateTime2literal
210         import pymysql.converters
211         pymysql.converters.conversions[Pendulum] = pymysql.converters.escape_datetime
212     except ImportError:
213         pass
```
that is solution.

**Comment by aa3222119** [comment]
@cdmikechen 确实可以解决，我的airflow.cfg配置的就是 mysql+pymysql:// ，建议这个可以提到版本里面，从配置中获取。

**Comment by eden-wang** [comment]
I try the method @cdmikechen provided and it worked. My environment is python 3.5.2, airflow 1.10.2(using pendulum 1.4.4), pymysql 0.9.3. I think it's the problem of airflow, not pendulum.

**Comment by lphcreat** [comment]
@cdmikechen it worked.,thanke you.

**Comment by ashb** [comment]
Thanks. We will fix this in Airflow 1.10.4
